presto-kafka/src/main/java/io/prestosql/plugin/kafka/KafkaRecordSet.java

```diff
@@ -269,4 +279,58 @@ public void close()
             kafkaConsumer.close();
         }
     }
+
+    public static FieldValueProvider headerMapValueProvider(MapType varcharMapType, Headers headers)
+    {
+        final Type keyType = varcharMapType.getTypeParameters().get(0);
+        final Type valueArrayType = varcharMapType.getTypeParameters().get(1);
+        final Type valueType = valueArrayType.getTypeParameters().get(0);
+
+        final BlockBuilder mapBlockBuilder = varcharMapType.createBlockBuilder(null, 1);
+        final BlockBuilder builder = mapBlockBuilder.beginBlockEntry();
+
+        // Group by keys and collect values as array.
+        final Map<String, List<byte[]>> headerMap = new HashMap<>();
+        for (Header header : headers) {
+            final String key = header.key();
+            if (headerMap.containsKey(key)) {
+                headerMap.get(key).add(header.value());
+            }
+            else {
+                final List<byte[]> list = new ArrayList<>();
+                list.add(header.value());
+                headerMap.put(key, list);
+            }
+        }
+
+        for (Map.Entry<String, List<byte[]>> header : headerMap.entrySet()) {
+            keyType.writeSlice(builder, utf8Slice(header.getKey()));
+            final BlockBuilder arrayBuilder = builder.beginBlockEntry();
+            for (byte[] value : header.getValue()) {
+                if (value == null) {
+                    arrayBuilder.appendNull();
+                }
+                else {
+                    valueType.writeSlice(arrayBuilder, wrappedBuffer(value));
+                }
+            }
+            builder.closeEntry();
+        }
+
+        mapBlockBuilder.closeEntry();
+
+        return new FieldValueProvider() {
+            @Override
+            public boolean isNull()
+            {
+                return !headers.iterator().hasNext();
```

Comment:
Is there a value in returning NULL when no headers?
Why not return empty map?

Reply:
I just liked `SELECT _headers FROM topic1 WHERE _headers IS NOT NULL;` better than `SELECT _headers FROM topic1 WHERE cardinality(_headers) > 0;` but thats not a strong opinion. In fact you are right that the headers are empty and not absent.